File: 백준/Silver/3184. 양/양.java
import java.io.*;
import java.util.*;

public class Main {
    static int r,c;
    static char[][] madang;
    static boolean[][]visited;
    static int[] dx = {0,0,-1,1};
    static int[] dy = {-1,1,0,0};
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        r = Integer.parseInt(st.nextToken());
        c = Integer.parseInt(st.nextToken());
        madang = new char[r][c];
        visited = new boolean[r][c];
        for(int i=0;i<r;i++){
           String line = br.readLine();
           for(int j=0;j<c;j++){
            madang[i][j] = line.charAt(j);
           }
        }
        int totalSheep = 0;
        int totalWolf = 0;
        for(int i=0;i<r;i++){
            for(int j=0;j<c;j++){
                if(!visited[i][j]&&madang[i][j]!='#'){
                    int[] res = bfs(i,j);
                    int sheep = res[0];
                    int wolf = res[1];
                    if(sheep>wolf){
                        totalSheep += sheep;
                    }else{
                        totalWolf += wolf;
                    }
                }
            }
        }
        System.out.println(totalSheep + " "  + totalWolf);
        
    }
    static int[] bfs(int startx, int starty){
        Queue<int[]> q = new LinkedList<>();
        q.offer(new int[] {startx,starty});
        visited[startx][starty] = true;

        int sheepcnt = 0;
        int wolfcnt = 0;
        if(madang[startx][starty]=='o'){
            sheepcnt++;
        }else if(madang[startx][starty]=='v'){
            wolfcnt++;
        }
        while(!q.isEmpty()){
            int[] cur = q.poll();
            int x = cur[0];
            int y = cur[1];
            for(int i=0;i<4;i++){
                int nx = x+dx[i];
                int ny = y+dy[i];
                if(nx<0 || nx>=r || ny<0 || ny>=c){
                    continue;
                }
                if(!visited[nx][ny] && madang[nx][ny]!='#'){
                    visited[nx][ny] = true;
                    q.offer(new int[]{nx,ny});
                    if(madang[nx][ny] =='o'){
                        sheepcnt++;
                    }else if(madang[nx][ny]=='v'){
                        wolfcnt++;
                    }
                }
            }
        }
        return new int[]{sheepcnt,wolfcnt};//양이랑 늑대 수 리턴..
    }
}

File: 백준/Silver/3184. 양/양.py
import sys
from collections import deque
input = sys.stdin.readline
r,c = map(int,input().split())
madang = [list(input().rstrip()) for _ in range(r)] #글자 '.' (점)은 빈 필드를 의미하며, 글자 '#'는 울타리를, 'o'는 양, 'v'는 늑대를 의미
visited = [[False]*c for _ in range(r)]
dx = [0, 0, 1, -1]
dy = [1, -1, 0, 0]
def bfs(x,y):
    q = deque([(x, y)])
    visited[y][x] = True
    sheep, wolf = 0,0
    if madang[y][x]=='o':
        sheep +=1
    elif madang[y][x] =='v':
        wolf+=1
    while q:
        x,y = q.popleft()
        for i in range(4):
            nx, ny = x + dx[i], y + dy[i]
            if 0 <= nx < c and 0 <= ny < r and not visited[ny][nx] and madang[ny][nx] != '#':

                q.append((nx,ny))
                visited[ny][nx] = True
                if madang[ny][nx]=='o':
                    sheep+=1
                elif madang[ny][nx]=='v':
                    wolf+=1
    return sheep, wolf
def solve():
    total_s, total_w = 0,0
    for x in range(c):
        for y in range(r):
            if madang[y][x]!= '#' and not visited[y][x]:
                sheep,wolf = bfs(x,y)
                if sheep>wolf:
                    total_s += sheep
                else:
                    total_w += wolf

    print(total_s, total_w)
solve()
